plugins/synaptics-rmi/fu-synaptics-rmi-device.c

```
@@ -0,0 +1,1070 @@
+			g_autofree FuSynapticsRmiFunction *func = NULL;
+			g_autoptr(GByteArray) res = NULL;
+			res = fu_synaptics_rmi_device_read (self, addr, RMI_DEVICE_PDT_ENTRY_SIZE, error);
+			if (res == NULL) {
+				g_prefix_error (error, "failed to read PDT entry @ 0x%04x: ", addr);
+				return FALSE;
+			}
+			func = fu_synaptics_rmi_function_parse (res, page_start, interrupt_count, error);
+			if (func == NULL)
+				return FALSE;
+			if (func->function_number == 0)
+				break;
+			interrupt_count += func->interrupt_source_count;
+			g_ptr_array_add (priv->functions, g_steal_pointer (&func));
+		}
+	}
+
+	/* see docs */
+	priv->num_interrupt_regs = (interrupt_count + 7) / 8;
+
+	/* for debug */
+	if (g_getenv ("FWUPD_SYNAPTICS_RMI_VERBOSE") != NULL) {
+		for (guint i = 0; i < priv->functions->len; i++) {
+			FuSynapticsRmiFunction *func = g_ptr_array_index (priv->functions, i);
+			g_debug ("PDT-%02u fn:0x%02x vr:%d sc:%d ms:0x%x "
+				 "db:0x%02x cb:0x%02x cm:0x%02x qb:0x%02x",
+				 i, func->function_number, func->function_version,
+				 func->interrupt_source_count,
+				 func->interrupt_mask,
+				 func->data_base,
+				 func->control_base, func->command_base,
+				 func->query_base);
+		}
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_synaptics_rmi_device_set_mode (FuSynapticsRmiDevice *self,
+				  FuSynapticsRmiHidMode mode,
+				  GError **error)
+{
+	FuSynapticsRmiDevicePrivate *priv = GET_PRIVATE (self);
+	const guint8 data[] = { 0x0f, mode };
+	gint rc;
+
+	fu_common_dump_raw (G_LOG_DOMAIN, "SetMode", data, sizeof(data));
+	rc = ioctl (priv->fd, HIDIOCSFEATURE(sizeof(data)), data);
+	if (rc < 0) {
+		if (rc == -EPERM) {
+			g_set_error_literal (error,
+					     FWUPD_ERROR,
+					     FWUPD_ERROR_PERMISSION_DENIED,
+					     "SetMode: permission denied");
+			return FALSE;
+		}
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "SetMode: ioctl failed with %s", strerror (rc));
+		return FALSE;
```

Is there really a lot of value to two different errors?  I would think strerror will get you a permission denied for -EPERM already.

---

Yup, I use the error code to know if we can ignore the error on detach, where the device node might have gone away.

---

Gotcha, makes sense.